DQMServices/Core/interface/locking_ptr.h

```diff
@@ -0,0 +1,136 @@
+#ifndef locking_ptr_h
+#define locking_ptr_h
+
+#include <memory>
+
+namespace detail {
+
+  class lock_base
+  {
+  public:
+    lock_base() = default;
+    virtual ~lock_base() = default;
+
+    virtual void lock() const = 0;
+    virtual void unlock() const = 0;
+  };
+
+  template <typename L>
+  class lock_impl : public lock_base
```

Comment:
That would work... though logically the lock type should be declared in the `MonitoerElement` class,

The reasoning was that the lock type depends on the implementation of the `MonitorElement` class (i.e. whether it needs a recursive lock or not), while the `locking_ptr` could stay generic enough for any lock type.

I can lose the generality, and make this into a more tightly coupled "THProxy" class that knows to look up the lock type in the `MonitorElement` class.